Added New Test and Troubleshoot the KubernetesPodOperator Locally Topic to Astro (#798)
astronomer/docs: astro/kubepodoperator-local.md

---
title: 'Test and Troubleshoot the KubernetesPodOperator Locally'
sidebar_label: 'KubernetesPodOperator (Local)'
id: kubepodoperator-local
description: Test and troubleshoot the KubernetesPodOperator locally.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import {siteVariables} from '@site/src/versions';

> The `KubernetesPodOperator` completes work within Kubernetes Pods in a Kubernetes cluster. Test the `KubernetesPodOperator` locally before running it in a production Kubernetes cluster. 

Author: @paolaperaza I have revised the introductory statement using some of the content you provided:

The `KubernetesPodOperator` is an Airflow operator that completes tasks in Kubernetes Pods. The `KubernetesPodOperator` provides an isolated, containerized execution environment for each task and lets you run custom Docker images and Python versions, set task-level resource requests, and more.

On Astro, the Kubernetes infrastructure required to run the `KubernetesPodOperator` is built in. To test the `KubernetesPodOperator` operator locally, you need a local Kubernetes environment.

> The `KubernetesPodOperator` completes work within Kubernetes Pods in a Kubernetes cluster. Test the `KubernetesPodOperator` locally before running it in a production Kubernetes cluster. 

Reviewer: Perfect

    user:
        client-certificate-data: <client-certificate-data>
        client-key-data: <client-key-data>
    ```
    The cluster `name` should be searchable as `docker-desktop` in your local `$HOME/.kube``config` file. Do not add any additional data to the `config` file.

2. Update the `<certificate-authority-data>`, `<certificate-authority-data>`, and `<certificate-authority-data>` values in the `config` file with the values for your organization. 
3. Under cluster, change `server: https://localhost:6445` to `server: https://kubernetes.docker.internal:6443` to identify the localhost running Kubernetes Pods. If this doesn't work, try `server: https://host.docker.internal:6445`.
4. Optional. Add the `.kube` folder to `.gitignore` if  your project is configureed as a GitHub repository and you want to prevent the file from being tracked by your version control tool. 
5. Optional. Add the `.kube` folder to `.dockerignore` to exclude it from the Docker image.

</TabItem>
<TabItem value="linux">

In a `.kube` folder in your Astro project, create a config file with:

```bash
microk8s.config > /include/.kube/config
```
</TabItem>
</Tabs>

## Step 3: Run your container

> To use the KubernetesPodOperator, you must define the configuration for your pod, including its namespace and runtime image, when you create an instance of the operator. This sample syntax runs the `hello-world` Docker image in a different namespace based on whether you're running the DAG locally or on Astro. If you are using Linux, the `cluster_context` is `microk8s`. The `config_file` points to the edited `/include/.kube/config` file.

Reviewer: ```suggestion
To use the KubernetesPodOperator, you must define the configuration of each task and the Kubernetes Pod in which it runs, including its namespace and Docker image.

This example DAG runs a `hello-world` Docker image. The namespace is determined dynamically based on whether you're running the DAG in your local environment or on Astro. If you are using Linux, the `cluster_context` is `microk8s`. The `config_file` points to the edited `/include/.kube/config` file.
```

I think the Docker image here actually does _not_ have to be a Runtime image -- that's sort of the magic of it. The Pod is still based on Runtime because the worker or scheduler is, but users can run arbitrary images in there

Run `astro dev start` to build this config into your image.

```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.configuration import conf
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'start_date': datetime(2019, 1, 1),
    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(minutes=5),
}

namespace = conf.get('kubernetes', 'NAMESPACE')

# This will detect the default namespace locally and read the

> # environment namespace when deployed to Astronomer.

Reviewer: ```suggestion
# environment namespace when deployed to Astro.
```

if namespace =='default':
    config_file = '/usr/local/airflow/include/.kube/config'
    in_cluster = False
else:
    in_cluster = True
    config_file = None

dag = DAG('example_kubernetes_pod', schedule_interval='@once', default_args=default_args)

with dag:
    KubernetesPodOperator(
        namespace=namespace,
        image="hello-world",
        labels={"foo": "bar"},
        name="airflow-test-pod",

>         task_id="task-one",

Author: @paolaperaza good catch. This has been corrected. Thanks!

        in_cluster=in_cluster,  # if set to true, will look in the cluster, if false, looks for file
        cluster_context="docker-desktop",  # is ignored when in_cluster is set to True
        config_file=config_file,
        is_delete_operator_pod=True,
        get_logs=True,
    )

```
## Step 4: View Kubernetes logs

> Optional. Review the logs for any pods that were created by the operator for issues.

Reviewer: ```suggestion
Optional. Review the logs for any Pods that were created by the operator for issues.
```

P sure we landed on `Pods` in style guide

<Tabs
    defaultValue="windows and mac"
    values={[
        {label: 'Windows and Mac', value: 'windows and mac'},
        {label: 'Linux', value: 'linux'},
    ]}>
<TabItem value="windows and mac">

> Run `kubectl get pods -n $namespace` or `kubectl logs {pod_name} -n $namespace` to examine the logs for the pod that just ran. By default, `docker-for-desktop` runs pods in the `default` namespace.

Reviewer: Should `kubectl` be a prerequisite? https://kubernetes.io/docs/reference/kubectl/

Author: @paolaperaza fixed.

> Run `kubectl get pods -n $namespace` or `kubectl logs {pod_name} -n $namespace` to examine the logs for the pod that just ran. By default, `docker-for-desktop` runs pods in the `default` namespace.

Reviewer: ```suggestion
Run `kubectl get pods -n $namespace` or `kubectl logs {pod_name} -n $namespace` to examine the logs for the Pod that just ran. By default, `docker-for-desktop` runs Pods in the `default` namespace.
```

---
title: 'Test and Troubleshoot the KubernetesPodOperator Locally'
sidebar_label: 'KubernetesPodOperator (Local)'
id: kubepodoperator-local
description: Test and troubleshoot the KubernetesPodOperator locally.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import {siteVariables} from '@site/src/versions';

The `KubernetesPodOperator` is an Airflow operator that completes tasks in Kubernetes Pods. The `KubernetesPodOperator` provides an isolated, containerized execution environment for each task and lets you run custom Docker images and Python versions, set task-level resource requests, and more.

On Astro, the Kubernetes infrastructure required to run the `KubernetesPodOperator` is built in. To test the `KubernetesPodOperator` operator locally, you need a local Kubernetes environment. 

## Step 1: Set up Kubernetes
<Tabs
    defaultValue="windows and mac"
    values={[
        {label: 'Windows and Mac', value: 'windows and mac'},
        {label: 'Linux', value: 'linux'},
    ]}>
<TabItem value="windows and mac">

> The latest versions of Docker for Windows and Mac let you run a single node Kubernetes cluster locally. If you are using Windows, see [Setting Up Docker for Windows and WSL to Work Flawlessly](https://nickjanetakis.com/blog/setting-up-docker-for-windows-and-wsl-to-work-flawlessly). If you are using Mac, see [Docker Desktop for Mac user manual](https://nickjanetakis.com/blog/setting-up-docker-for-windows-and-wsl-to-work-flawlessly). It isn't nevessary to install docker-compose. 

Reviewer: ```suggestion
The latest versions of Docker for Windows and Mac let you run a single node Kubernetes cluster locally. If you are using Windows, see [Setting Up Docker for Windows and WSL to Work Flawlessly](https://docs.docker.com/desktop/windows/install/). If you are using Mac, see [Docker Desktop for Mac user manual](https://docs.docker.com/desktop/mac/). It isn't necessary to install [Docker Compose](https://docs.docker.com/compose/).
```

- I think `Docker Compose` is proper noun? https://docs.docker.com/compose/
- We prob shouldn't link to some person's personal blog post from 2020. Tried to find the right links from Docker docs but you can double check me on these!

1. Open Docker and go to **Settings** > **Kubernetes**.

2. Select the `Enable Kubernetes` checkbox. 

3. Click **Apply and Restart**.

4. Click **Install** in the **Kubernetes Cluster Installation** dialog.

    Docker restarts and the status indicator changes to green to indicate Kubernetes is running.

</TabItem>
<TabItem value="linux">

1. Install Microk8s. See [Microk8s](https://microk8s.io/).

2. Run `microk8s.start` to start Kubernetes.

</TabItem>
</Tabs>

## Step 2: Update the kubeconfig file

<Tabs
    defaultValue="windows and mac"
    values={[
        {label: 'Windows and Mac', value: 'windows and mac'},
        {label: 'Linux', value: 'linux'},
    ]}>
<TabItem value="windows and mac">

> 1. Go to the `$HOME/.kube` directory that was created when you enabled Kubernetes in Docker and copy the `config` file into the `/include/.kube/` folder in your Astro project. The `config` file contains all the information the KubePodOperator uses to connect to your cluster. For example:

Reviewer: ```suggestion
1. Go to the `$HOME/.kube` directory that was created when you enabled Kubernetes in Docker and copy the `config` file into the `/include/.kube/` folder in your Astro project. The `config` file contains all the information the KubernetesPodOperator uses to connect to your cluster. For example:
```
```suggestion
1. Go to the `$HOME/.kube` directory that was created when you enabled Kubernetes in Docker and copy the `config` file into the `/include/.kube/` folder in your Astro project. The `config` file contains all the information the KubePodOperator uses to connect to your cluster. For example:
```

3. Under cluster, change `server: https://localhost:6445` to `server: https://kubernetes.docker.internal:6443` to identify the localhost running Kubernetes Pods. If this doesn't work, try `server: https://host.docker.internal:6445`.
4. Optional. Add the `.kube` folder to `.gitignore` if  your Astro project is hosted in a GitHub repository and you want to prevent the file from being tracked by your version control tool.
5. Optional. Add the `.kube` folder to `.dockerignore` to exclude it from the Docker image.

</TabItem>
<TabItem value="linux">

In a `.kube` folder in your Astro project, create a config file with:

```bash
microk8s.config > /include/.kube/config
```
</TabItem>
</Tabs>

## Step 3: Run your container

To use the KubernetesPodOperator, you must define the configuration of each task and the Kubernetes Pod in which it runs, including its namespace and Docker image.

This example DAG runs a `hello-world` Docker image. The namespace is determined dynamically based on whether you're running the DAG in your local environment or on Astro. If you are using Linux, the `cluster_context` is `microk8s`. The `config_file` points to the edited `/include/.kube/config` file.

> Run `astro dev start` to build this config into your image.

Reviewer: ```suggestion
Once you've updated the definition of KubernetesPodOperator tasks in your Astro project, run `astro dev start` with the Astro CLI to test your DAGs in a local Airflow environment.
```

```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.configuration import conf
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,

>     'start_date': datetime(2019, 1, 1),

Reviewer: ```suggestion
    'start_date': datetime(2022, 1, 1),
```

    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 1,
    'retry_delay': timedelta(minutes=5),
}

namespace = conf.get('kubernetes', 'NAMESPACE')

# This will detect the default namespace locally and read the
# environment namespace when deployed to Astro.
if namespace =='default':
    config_file = '/usr/local/airflow/include/.kube/config'
    in_cluster = False
else:
    in_cluster = True
    config_file = None

dag = DAG('example_kubernetes_pod', schedule_interval='@once', default_args=default_args)

with dag:
    KubernetesPodOperator(
        namespace=namespace,
        image="hello-world",
        labels={"<pod-label>": "<label-name>"},
        name="airflow-test-pod",

>         task_id="task-one",

Reviewer: placeholder values?

Author: These are the pre-defined values for the sample DAG. A user doesn't change these values.

        in_cluster=in_cluster,  # if set to true, will look in the cluster, if false, looks for file
        cluster_context="docker-desktop",  # is ignored when in_cluster is set to True
        config_file=config_file,
        is_delete_operator_pod=True,
        get_logs=True,
    )

```
## Step 4: View Kubernetes logs

> Optional. Use the `kubectl` command line tool to review the logs for any pods that were created by the operator for issues. If you haven't installed the `kubectl` command line tool, see [Install Tools](https://kubernetes.io/docs/tasks/tools/#kubectl).

Reviewer: ```suggestion
Optional. Use the `kubectl` command line tool to review the logs for any Pods that were created by the operator for issues and help with troubleshooting. If you haven't installed the `kubectl` command line tool, see [Install Tools](https://kubernetes.io/docs/tasks/tools/#kubectl).
```

---
title: 'Test and Troubleshoot the KubernetesPodOperator Locally'
sidebar_label: 'Local KubernetesPodOperator'
id: kubepodoperator-local
description: Test and troubleshoot the KubernetesPodOperator locally.
---

You use the `KubernetesPodOperator` to create and run Kubernetes pods on a  Kubernetes cluster in a local Kubernetes environment. Using the `KubernetesPodOperator` locally allows you to create, test, and modify your code before you move to production.

## Set up Kubernetes

> 

Author: @jwitz I was reviewing the content today and had the same thought. :)

I have added tabs where I think it is necessary.

### Windows and Mac

> The latest versions of Docker for Windows and Mac let you run a single node Kubernetes cluster locally. If you are using Windows, see [Setting Up Docker for Windows and WSL to Work Flawlessly](https://nickjanetakis.com/blog/setting-up-docker-for-windows-and-wsl-to-work-flawlessly). If you are using Mac, see [Docker Desktop for Mac user manual](https://nickjanetakis.com/blog/setting-up-docker-for-windows-and-wsl-to-work-flawlessly). It isn't nevessary to install docker-compose.

Author: @jwitz these aren't prerequisites. They are links to supplementary information. 

There are two versions of Docker being discussed; Windows and Mac. Therefore, _version_ should be plural (versions).  

Here is the modified text:

_The latest versions of Docker for Windows and Mac both support running a single-node Kubernetes cluster locally._

1. Open Docker and go to **Settings** > **Kubernetes**.

2. Select the `Enable Kubernetes` checkbox. 

3. Click **Apply and Restart**.

4. Click **Install** in the **Kubernetes Cluster Installation** dialog.

    Docker restarts and the status indicator changes to green to indicate Kubernetes is running.

### Linux

> 1. Install Microk8s. See [Microk8s](https://microk8s.io/).

Author: @jwitz as there is only a single prerequisite, it is better to incorporate it into the task. One less heading.

2. Run `microk8s.start` to start Kubernetes.

## Update the kubeconfig File

### Windows and Mac

1. Go to the `$HOME/.kube` directory that was created when you enabled Kubernetes in Docker and copy the `config` file into the `/include/.kube/` folder in your Astro project. The `config` file contains all the information the KubePodOperator uses to connect to your cluster. For example:
    ```apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: CERT_AUTHORITY_DATA
        server: https://kubernetes.docker.internal:6443/
    name: docker-desktop
    contexts:
    - context:
        cluster: docker-desktop
        user: docker-desktop
    name: docker-desktop
    current-context: docker-desktop
    kind: Config
    preferences: {}
    users:
    - name: docker-desktop
    user:
        client-certificate-data: CERT_CLIENT_DATA
        client-key-data: CERT_KEY_DATA
    ```
    The cluster `name` should be searchable as `docker-desktop` in your local `$HOME/.kube``config` file. Do not add any additional data to the `config` file.

> 2. Update the `CERT_AUTHORITY_DATA`, `CERT_CLIENT_DATA`, and `CERT_KEY_DATA` values in the `config` file with the values for your organization.

Author: @jwitz is there an edit you're recommending?

> 2. Update the `CERT_AUTHORITY_DATA`, `CERT_CLIENT_DATA`, and `CERT_KEY_DATA` values in the `config` file with the values for your organization.

Reviewer 2: Whoops not sure why this suggestion didn't go through! I recommended using `<cert-authority-data>` etc to match our placeholder value styling. Currently they look like system environment variables. Just updated the suggestion to reflect that. If you do make this change, these values should also change in the code block itself.

3. Under cluster, change `server: https://localhost:6445` to `server: https://kubernetes.docker.internal:6443` to identify the localhost running Kubernetes Pods. If this doesn't work, try `server: https://host.docker.internal:6445`.
4. Optional. Add the `.kube` folder to `.gitignore` if your project is configureed as a GitHub repository and you want to prevent the file from being tracked by your version control tool.
5. Optional. Add the `.kube` folder to `.dockerignore` to exclude it from the Docker image.

### Linux

In a `.kube` folder in your Astro project, create a config file with:

```bash
microk8s.config > /include/.kube/config
```

## Run Your Container

> The `config_file` points to the `/include/.kube/config` file you just edited. 

Reviewer 2: I'm cool with "create an instance"!

In your alternatives, the namespace/ configuration is part of the example explanation, but I want that to be a part of what we instruct the user to do as a requirement whenever they run KPO. So like:

`To use the KubernetesPodOperator, you must define the configuration for your pod, including its namespace and runtime image, when you create an instance of the operator. This sample syntax runs the `hello-world` Docker image in a different namespace based on whether you're running the DAG locally or on Astro.`

The following example runs the docker `hello-world` image and reads environment variables to determine where it is run. If you are using Linux, the `cluster_context` is `microk8s`.

> Run `astro dev start` to build this config into your image.

Author: @jwitz I think this information should remain with the example syntax.

    'retries': 1,
    'retry_delay': timedelta(minutes=5),
}

namespace = conf.get('kubernetes', 'NAMESPACE')

# This will detect the default namespace locally and read the
# environment namespace when deployed to Astronomer.
if namespace =='default':
    config_file = '/usr/local/airflow/include/.kube/config'
    in_cluster = False
else:
    in_cluster = True
    config_file = None

dag = DAG('example_kubernetes_pod', schedule_interval='@once', default_args=default_args)

with dag:
    KubernetesPodOperator(
        namespace=namespace,
        image="hello-world",
        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster,  # when set to True, it looks in the cluster, if False, it looks for a file
        cluster_context="docker-desktop",  # is ignored when in_cluster is set to True
        config_file=config_file,
        get_logs=True,
    )

```
## View Kubernetes Logs

### Windows and Mac

Run `kubectl get pods -n $namespace` or `kubectl logs {pod_name} -n $namespace` to examine the logs for the pod that just ran. By default, `docker-for-desktop` and `microk8s` run pods in the `default` namespace.

> ### Linux

Author: @jwitz done. Take a look and provide your feedback.

Run the same commands prefixed with microk8s. For example:
```
microk8s.kubectl get pods -n $namespace
```

> ## Test the KubernetesPodOperator Locally

Reviewer 2: I'm all for having islands when it saves you from jumping between different documents, but repeating the same information and code sample in a single doc and linking internally feels less intuitive. Really, the only differences for this content are that the example is code is less detailed and we don't put `Step 1/2/3` for the first topics. 

What's the benefit of having this duplicate section instead of organizing the entire document as a procedure? I'd prefer doing the latter by removing this topic and framing the other topics as steps. Because a user is likely to only need to "Set up Kubernetes" and "Update the kubeconfig file" once per environment, I don't think it's necessary to frame that topic as a general documentation outside of a procedure.

        namespace=namespace,
        image="hello-world",
        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster,  # when set to True, it looks in the cluster, if False, it looks for a file
        cluster_context="docker-desktop",  # is ignored when in_cluster is set to True
        config_file=config_file,
        is_delete_operator_pod=True,
        get_logs=True,
    )

```
## View Kubernetes Logs

### Windows and Mac

Run `kubectl get pods -n $namespace` or `kubectl logs {pod_name} -n $namespace` to examine the logs for the pod that just ran. By default, `docker-for-desktop` and `microk8s` run pods in the `default` namespace.

### Linux

Run the same commands prefixed with microk8s. For example:
```
microk8s.kubectl get pods -n $namespace
```
## Test the KubernetesPodOperator Locally

Testing DAGs with the [KubernetesPodOperator](kubernetespodoperator.md) locally requires a local Kubernetes environment.

### Step 1: Prepare Your Environment

1. Set up Kubernetes in Docker. See [Set up Kubernetes](#set-up-kubernetes).
2. Update the kubeconfig File. See [Update the kubeconfig File](#update-the-kubeconfig-file).

> ### Step 2: Instantiate the KubernetesPodOperator

Author: @dstandish can you clarify your concern? Are you requesting the removal of  _Add a KubernetesPodOperator Task_ or  _Run Your Container_?

> ### Step 2: Instantiate the KubernetesPodOperator

Reviewer 3: no concern, do what you like, just from quick glance seemed like you were repeating an example

To instantiate the KubernetesPodOperator in a given DAG, update your DAG file to include the following code:

```python
from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
from airflow import configuration as conf
# ...

namespace = conf.get('kubernetes', 'NAMESPACE')

# This will detect the default namespace locally and read the
# environment namespace when deployed to Astronomer.
if namespace =='default':
    config_file = '/usr/local/airflow/include/.kube/config'
    in_cluster=False
else:
    in_cluster=True
    config_file=None

with dag:
    k = KubernetesPodOperator(
        namespace=namespace,
        image="my-image",
        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster, # if set to true, will look in the cluster for configuration. if false, looks for file
        cluster_context='docker-desktop', # is ignored when in_cluster is set to True
        config_file=config_file,

>         is_delete_operator_pod=True,

Reviewer 3: ```suggestion
```

---
title: 'Test and Troubleshoot the KubernetesPodOperator Locally'

> sidebar_label: 'Run the KubernetesPodOperator Locally'

Author: @jwitz I don't think this change is necessary.

> sidebar_label: 'Run the KubernetesPodOperator Locally'

Author: @jwitz I implemented these changes.

id: kubepodoperator-local
description: Test and troubleshoot the KubernetesPodOperator locally.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import {siteVariables} from '@site/src/versions';

The `KubernetesPodOperator` completes work within Kubernetes Pods in a Kubernetes cluster. Test the `KubernetesPodOperator` locally before running it in a production Kubernetes cluster. 

> ## Set up Kubernetes

Reviewer 2: ```suggestion
## Step 1: Set up Kubernetes
```
See my comment later on about this suggestion. This is an example of what I'm talking about.

<Tabs
    defaultValue="windows and mac"
    values={[
        {label: 'Windows and Mac', value: 'windows and mac'},
        {label: 'Linux', value: 'linux'},
    ]}>

> <TabItem value="windows and mac">

Author: @jwitz the topics have been updated and now are identical (save a few minor differences).

<TabItem value="linux">

1. Install Microk8s. See [Microk8s](https://microk8s.io/).

2. Run `microk8s.start` to start Kubernetes.

</TabItem>
</Tabs>

## Update the kubeconfig File
<Tabs
    defaultValue="windows and mac"
    values={[
        {label: 'Windows and Mac', value: 'windows and mac'},
        {label: 'Linux', value: 'linux'},
    ]}>
<TabItem value="windows and mac">

1. Go to the `$HOME/.kube` directory that was created when you enabled Kubernetes in Docker and copy the `config` file into the `/include/.kube/` folder in your Astro project. The `config` file contains all the information the KubePodOperator uses to connect to your cluster. For example:
    ```apiVersion: v1
    clusters:
    - cluster:
        certificate-authority-data: <cert-authority-data>
        server: https://kubernetes.docker.internal:6443/
    name: docker-desktop
    contexts:
    - context:
        cluster: docker-desktop
        user: docker-desktop
    name: docker-desktop
    current-context: docker-desktop
    kind: Config
    preferences: {}
    users:
    - name: docker-desktop
    user:
        client-certificate-data: CERT_CLIENT_DATA
        client-key-data: CERT_KEY_DATA
    ```

>     The `clusters.cluster.name` should be searchable as `docker-desktop` in your local `$HOME/.kube``config` file.

Reviewer 2: ```suggestion
    The `clusters.cluster.name` should be searchable as `docker-desktop` in your local `$HOME/.kube/config` file.
```

2. Update the `CERT_AUTHORITY_DATA`, `CERT_CLIENT_DATA`, and `CERT_KEY_DATA` values in the `config` file with the values for your organization.
3. Under cluster, change `server: https://localhost:6445` to `server: https://kubernetes.docker.internal:6443` to identify the localhost running Kubernetes Pods. If this doesn't work, try `server: https://host.docker.internal:6445`.
4. Optional. Add the `.kube` folder to `.gitignore` if your project is configureed as a GitHub repository and you want to prevent the file from being tracked by your version control tool.
5. Optional. Add the `.kube` folder to `.dockerignore` to exclude it from the Docker image.

</TabItem>

<TabItem value="linux">

In a `.kube` folder in your Astro project, create a default `config` file with:

```bash
microk8s config > /include/.kube/config
```
</TabItem>
</Tabs>

> ## Run Your Container

Author: @jwitz the topics have been updated and now are identical (save a few minor differences).

The `config_file` points to the `/include/.kube/config` file you just edited. 

> You can use the following example DAG to test your configurations. This DAG runs the `hello-world` Docker image in a different cluster based on whether you're running the DAG locally or on Astro. 

Reviewer 2: ```suggestion
You can use the following example DAG to test your configurations. This DAG runs the `hello-world` Docker image in a different namespace based on whether you're running the DAG locally or on Astro. 
```

Author: @jwitz with the previous edit, this text has been removed.

</TabItem>
</Tabs>

## Test the KubernetesPodOperator Locally

Testing DAGs with the [KubernetesPodOperator](kubernetespodoperator.md) locally requires a local Kubernetes environment.

### Step 1: Prepare Your Environment

1. Set up Kubernetes in Docker. See [Set up Kubernetes](#set-up-kubernetes).
2. Update the kubeconfig file. See [Update the kubeconfig File](#update-the-kubeconfig-file).

### Step 2: Add a KubernetesPodOperator Task

To add the KubernetesPodOperator task in a DAG, update your DAG file to include the following code:

```python
from airflow.contrib.operators.kubernetes_pod_operator import KubernetesPodOperator
from airflow import configuration as conf
# ...

namespace = conf.get('kubernetes', 'NAMESPACE')

# This will detect the default namespace locally and read the
# environment namespace when deployed to Astronomer.
if namespace =='default':
    config_file = '/usr/local/airflow/include/.kube/config'
    in_cluster=False
else:
    in_cluster=True
    config_file=None

with dag:
    k = KubernetesPodOperator(
        namespace=namespace,

>         image="my-image",

Reviewer 2: These are all placeholder values, but we don't really describe them as such. We should either use our normal placeholder syntax or have more documentation about what to fill in here.

        labels={"foo": "bar"},
        name="airflow-test-pod",
        task_id="task-one",
        in_cluster=in_cluster, # if set to true, will look in the cluster for configuration. if false, looks for file
        cluster_context='docker-desktop', # is ignored when in_cluster is set to True
        config_file=config_file,
        get_logs=True)
```

Specifically, your operator must have `cluster_context='docker-desktop` and `config_file=config_file`.

### Step 3: Run and Monitor the KubernetesPodOperator

1. Run `astro dev restart` in the Astro CLI to rebuild your image and run your project in a local Airflow environment.
2. Optional. Review the logs for any pods that were created by the operator for issues. See [View Kubernetes Logs](#view-kubernetes-logs).

> ## Next Steps ##

Reviewer 2: ```suggestion
## Next Steps
```
